### PR TITLE
Remove public fields of BlueskyContext and replace with methods

### PR DESCRIPTION
**Pull request by callumforrester**
It seems more sustainable to use private fields and access them via methods. The device and plan dictionaries and run engine are now hidden behind methods.

### CODE REVIEW

File: src/blueapi/core/context.py [hunk 1]
```diff
@@ -37,21 +38,44 @@
...
+    _devices: Dict[str, Device]
+    _plan_functions: Dict[str, PlanGenerator]
 
     _reference_cache: Dict[Type, Type] = field(default_factory=dict)
 
+    def __init__(self, run_engine: Optional[RunEngine] = None) -> None:
+        if run_engine is None:
+            run_engine = RunEngine(context_managers=[])
+
+        self._run_engine = run_engine
```
Comment (tpoliaw) [inline src/blueapi/core/context.py:55]: `self._run_engine = run_engine or RunEngine(context_managers=[])`?

File: src/blueapi/core/context.py [hunk 2]
```diff
@@ -37,21 +38,44 @@
...
-    )
-    plans: Dict[str, Plan] = field(default_factory=dict)
-    devices: Dict[str, Device] = field(default_factory=dict)
-    plan_functions: Dict[str, PlanGenerator] = field(default_factory=dict)
+    _run_engine: RunEngine
+    _plans: Dict[str, Plan]
+    _devices: Dict[str, Device]
+    _plan_functions: Dict[str, PlanGenerator]
 
     _reference_cache: Dict[Type, Type] = field(default_factory=dict)
```
Comment (tpoliaw) [inline src/blueapi/core/context.py]: Shouldn't this field be removed as well?

File: src/blueapi/core/context.py [hunk 3]
```diff
@@ -37,21 +38,44 @@
...
+    def run_engine(self) -> RunEngine:
+        return self._run_engine
+
+    def find_plan_function(self, name: str) -> Optional[PlanGenerator]:
+        return self._plan_functions.get(name)
+
+    def find_plan_metadata(self, name: str) -> Optional[Plan]:
+        return self._plans.get(name)
+
+    def all_devices(self) -> Iterable[Device]:
```
Comment (tpoliaw) [inline src/blueapi/core/context.py:100]: worth making these two properties as well?

File: src/blueapi/worker/task.py [hunk 4]
```diff
@@ -41,15 +41,18 @@ class RunPlan(Task):
     def do_task(self, ctx: BlueskyContext) -> None:
         LOGGER.info(f"Asked to run plan {self.name} with {self.params}")
 
-        plan = ctx.plans[self.name]
-        func = ctx.plan_functions[self.name]
-        sanitized_params = _lookup_params(ctx, plan, self.params)
-        plan_generator = func(**sanitized_params.dict())
+        metadata = ctx.find_plan_metadata(self.name)
```
Comment (tpoliaw) [inline src/blueapi/worker/task.py:44]: If these two functions are only ever called together and both have to be present, it'd be clearer to have a single `find_plan(name) -> Optional[(Plan, PlanFunc)]`. Probably suggests `plans` and `plan_functions` should be merged into a single field in the context as well. Maybe a dict of `name -> (Plan, func)`?

**Review by tpoliaw — COMMENTED** [review]
There are a few docs changes here not related to replacing fields with methods. Is it picky to suggest these are committed/merged as separate commits when the PR is merged?

File: src/blueapi/core/context.py (hunk 2 above)
Comment (callumforrester) [reply src/blueapi/core/context.py]: Why? It's in use and already private

Comment (tpoliaw) [reply src/blueapi/core/context.py]: I meant the `= field(default_factory=dict)` part, but it looks like you changed it between me writing and submitting the comment

**Comment by callumforrester** [comment]
Very out of date, we may revisit what to do about `BlueskyContext` in future, it has become quite large and bloated... https://github.com/DiamondLightSource/blueapi/issues/616

**Pull request closed without merging**